symengine/utilities/teuchos/Teuchos_RCPNode.cpp

```diff
@@ -52,6 +52,11 @@
 // Internal implementatation stuff
 //
 
+#if defined(TEUCHOS_DEBUG) && defined(WITH_SYMENGINE_THREAD_SAFE)
```

Comment:
This is backported and modified to use `WITH_SYMENGINE_THREAD_SAFE`, taken from here:
https://github.com/trilinos/Trilinos/blob/293bc60e0456a04fbeaad5505df6a8a08084573f/packages/teuchos/core/src/Teuchos_RCPNode.cpp#L27

The rest of the changes in this file can be found by simply searching for "mutex" throughout the file.

This should hopefully resolve the data race (https://github.com/symengine/symengine/issues/2114#issue-3229206443) without requiring an invasive full update of Teuchos RCP as in #2115